sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +129,49 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    // As the analyzer makes sure that the list of In is already of the same data type, then the
+    // rule can simply check the first literal in `in.list` can implicitly cast to `toType` or not,
+    // and this rule doesn't convert in when `in.list` is empty.
+    case in @ In(Cast(fromExp, toType: NumericType, _), list @ Seq(firstLit, _*))
+        if canImplicitlyCast(fromExp, toType, firstLit.dataType) && in.inSetConvertible =>
+      val (newValueList, exp) =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .partition {
+            case EqualTo(_, _: Literal) => true
+            case _ => false
+          }
+
+      val (nonNullValueList, nullValueList) = newValueList.partition {
+        case EqualTo(_, NonNullLiteral(_, _: NumericType)) => true
+        case _ => false
+      }
+      // make sure the new return list have the same dataType.
+      val newList = {
+        if (nonNullValueList.nonEmpty) {
+          // cast the null value to the dataType of nonNullValueList
+          // when the nonNullValueList is nonEmpty.
+          nullValueList.map {
+            case EqualTo(_, lit) =>
+              Cast(lit, nonNullValueList.head.asInstanceOf[EqualTo].left.dataType)
+          } ++ nonNullValueList.map {case EqualTo(_, lit) => lit}
+        } else {
+          // the new value list only contains null value,
+          // cast the null value to fromExp.dataType.
+          nullValueList.map {
+            case EqualTo(_, lit) =>
+              Cast(lit, fromExp.dataType)
+          }
+        }
+      }
+
+      val unwrapIn = In(fromExp, newList)
+      // since `exp` are all the same,
+      // convert to a single value `And(IsNull(_), Literal(null, BooleanType))`.
+      exp.headOption match {
```

Comment:
we can assume `assert(exp.length == 1, ...)`?

Reply:
Actually, the `expr.lenght >= 1` when the `expr` is nonEmpty, as @sunchao mention [above](https://github.com/apache/spark/pull/32488#discussion_r631497573), we convert `expr` to a single value because the values in `expr` are all the same.

Reply:
If so, `assert(expr.distinct.length == 1, ...)` or `assert(expr.map(_.canonicalized).distinct.length == 1, ...)` instead? Anyway, I think it is better to check if `expr` has an expected state.

Reply:
SGTM, will do.

Reply:
Add new condition `expr.map(_.canonicalized).distinct.length == 1` when the `expr` is nonEmpty, return original expression when `expr` has an unexpected state.